util/election2/sql/election.go

```diff
@@ -0,0 +1,269 @@
+func (e *Election) becomeLeaderLoop(ctx context.Context, closer chan *chan error) {
+	for {
+		select {
+		case ch := <-closer:
+			err := e.tearDown()
+			klog.Infof("Election teardown for %s: %v", e.resourceID, err)
+			if ch != nil {
+				*ch <- err
+			}
+			return
+		default:
+			leader, err := e.tryBecomeLeader(ctx)
+			if err != nil {
+				klog.Errorf("Failed attempt to become leader for %s, retrying: %v", e.resourceID, err)
+			} else {
+				e.leaderLock.L.Lock()
+				if leader != e.currentLeader {
+					// Note: this code does not actually care _which_ instance was
+					// elected, it sends notifications on each leadership cahnge.
+					e.currentLeader = leader
+					e.leaderLock.Broadcast()
+				}
+				e.leaderLock.L.Unlock()
+			}
+			time.Sleep(e.electionInterval)
+		}
+	}
+}
+
+func (e *Election) tryBecomeLeader(ctx context.Context) (leaderData, error) {
+	leader := leaderData{}
+	tx, err := e.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return leader, fmt.Errorf("BeginTX: %w", err)
+	}
+	defer tx.Rollback()
+	row := tx.QueryRow(
+		"SELECT leader, last_update FROM leader_election WHERE resource_id = $1",
+		e.resourceID)
+	if err := row.Scan(&leader.currentLeader, &leader.timestamp); err != nil {
+		return leader, fmt.Errorf("Select: %w", err)
+	}
+	
+	if leader.currentLeader != e.instanceID && leader.timestamp.Add(e.electionInterval*10).After(time.Now()) {
+		return leader, nil // Someone else won the election
+	}
+
+	timestamp := time.Now()
+	_, err = tx.Exec(
+		"UPDATE leader_election SET leader = $1, last_update = $2 WHERE resource_id = $3 AND leader = $4 AND last_update = $5",
+		e.instanceID, timestamp, e.resourceID, leader.currentLeader, leader.timestamp)
+	if err != nil {
+		return leader, fmt.Errorf("Update: %w", err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return leader, fmt.Errorf("Commit failed: %w", err)
+	}
+	leader = leaderData{currentLeader: e.instanceID, timestamp: timestamp}
+	return leader, nil
+}
+
+func (e *Election) tearDown() error {
+	e.leaderLock.L.Lock()
+	defer e.leaderLock.L.Unlock()
+	if e.currentLeader.currentLeader != e.instanceID {
+		return nil
+	}
+	e.currentLeader.currentLeader = "empty leader"
+	e.leaderLock.Broadcast()
+
+	// Reset election time to epoch to allow a faster fail-over
+	res, err := e.db.Exec(
+		"UPDATE leader_election SET last_update = $1 WHERE resource_id = $2 AND leader = $3 AND last_update = $4",
+		time.Time{}, e.resourceID, e.instanceID, e.currentLeader.timestamp)
+	if err != nil {
+		return fmt.Errorf("Update: %w", err)
+	}
+	if n, err := res.RowsAffected(); n != 1 || err != nil {
+		return fmt.Errorf("failed to resign leadership: %d, %w", n, err)
+	}
+	return nil
+}
+
+func (e *Election) initializeLock(ctx context.Context) error {
+	insert, err := e.db.Prepare("INSERT INTO leader_election (resource_id, leader, last_update) VALUES ($1, $2, $3)")
```

Comment:
Wouldn't we need to handle the case where more than one replica executes an initialization at the same time? Or is the idea is to crash loop, restart and continue?